sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/util/S3Util.java

```diff
@@ -0,0 +1,573 @@
+              .withDestinationKey(destinationPath.getKey())
+              .withUploadId(uploadId)
+              .withPartNumber(partNumber)
+              .withFirstByte(bytePosition)
+              .withLastByte(Math.min(objectSize - 1, bytePosition + s3UploadBufferSizeBytes - 1));
+
+      tasks.add(
+          new Callable<PartETag>() {
+            @Override
+            public PartETag call() throws IOException {
+              CopyPartResult copyPartResult;
+              try {
+                copyPartResult = amazonS3.copyPart(copyPartRequest);
+              } catch (AmazonClientException e) {
+                throw new IOException(e);
+              }
+              return copyPartResult.getPartETag();
+            }
+          });
+
+      bytePosition += s3UploadBufferSizeBytes;
+    }
+
+    ExecutorService executorService;
+    if (tasks.size() > 1) {
+      // Don't pollute the main thread pool, which this was called from.
+      // Instead, create a small thread pool just for this copy operation.
+      int threadQuantity = Math.min(MAX_THREADS_PER_CONCURRENT_COPY, tasks.size());
+      executorService = Executors.newFixedThreadPool(threadQuantity);
+    } else {
+      // Don't create a thread pool if there is only one task.
+      executorService = MoreExecutors.newDirectExecutorService();
+    }
+    List<PartETag> eTags;
+    try {
+      eTags = invokeAllAndUnwrapResults(tasks, executorService);
+    } finally {
+      executorService.shutdown();
+    }
+
+    CompleteMultipartUploadRequest completeUploadRequest =
+        new CompleteMultipartUploadRequest()
+            .withBucketName(destinationPath.getBucket())
+            .withKey(destinationPath.getKey())
+            .withUploadId(uploadId)
+            .withPartETags(eTags);
+
+    try {
+      amazonS3.completeMultipartUpload(completeUploadRequest);
+    } catch (AmazonClientException e) {
+      throw new IOException(e);
+    }
+  }
+
+  public void delete(List<S3Path> paths) throws IOException {
+    Multimap<String, String> keysByBucket = ArrayListMultimap.create();
+    for (S3Path path : paths) {
+      keysByBucket.put(path.getBucket(), path.getKey());
+    }
+
+    List<Callable<Void>> tasks = new ArrayList<>();
+    for (final String bucket : keysByBucket.keySet()) {
+      for (final List<String> keysPartition : Iterables.partition(keysByBucket.get(bucket), 1000)) {
+        tasks.add(
+            new Callable<Void>() {
+              @Override
+              public Void call() throws IOException {
+                delete(bucket, keysPartition);
+                return null;
+              }
+            });
+      }
+    }
+
+    invokeAllAndUnwrapResults(tasks, executorService);
+  }
+
+  private void delete(String bucket, Collection<String> keys) throws IOException {
+    // S3 SDK: "You may specify up to 1000 keys."
+    checkArgument(keys.size() <= 1000, "only 1000 keys can be deleted per request");
```

Comment:
Is this a limitation that propagates into Beam, or do you have chunking into 1000-key batches somewhere at a higher level?

Reply:
The method is private, and the only usage is a few lines away. I'll change this to a constant value to make this more readable and less error-prone.

Reply:
Done